ast-visual-studio-extension/CxExtension/Services/ASCAService.cs

```diff
@@ -0,0 +1,490 @@
+                if (document != null)
+                {
+
+                    // Retrieve document content
+                    var textDocument = (TextDocument)document.Object("TextDocument");
+                    var content = textDocument.StartPoint.CreateEditPoint().GetText(textDocument.EndPoint);
+
+                    // Use the same file name but save it in the Temp directory
+                    var originalFileName = Path.GetFileName(document.FullName);
+                    var tempFilePath = Path.Combine(Path.GetTempPath(), originalFileName);
+                    File.WriteAllText(tempFilePath, content);
+                    Debug.WriteLine($"Temporary file created: {tempFilePath}");
+
+                    try
+                    {
+                        WriteToOutputPane($"Start ASCA scan On File: {document.FullName}");
+
+                        // Perform ASCA scan
+                        CxAsca scanResult = await cxWrapper.ScanAscaAsync(
+                            fileSource: tempFilePath,
+                            ascaLatestVersion: false,
+                            agent: "Visual Studio"
+                        );
+
+                        if (scanResult.Error != null)
+                        {
+                            Debug.WriteLine($"ASCA scan failed: {scanResult.Error.Description}");
+                            return;
+                        }
+
+                        Debug.WriteLine("ASCA scan completed successfully.");
+                        await DisplayDiagnosticsAsync(scanResult.ScanDetails, document.FullName);
+                    }
+                    finally
+                    {
+                        // Delete the temporary file
+                        if (File.Exists(tempFilePath))
+                        {
+                            File.Delete(tempFilePath);
+                            Debug.WriteLine($"Temporary file deleted: {tempFilePath}");
+                        }
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Debug.WriteLine($"Failed to process document: {ex.Message}");
+            }
+        }
+
+        private void ClearAllMarkers()
+        {
+            foreach (var marker in _activeMarkers)
+            {
+                try
+                {
+                    marker.Invalidate(); // Removes the marker
+                }
+                catch (Exception ex)
+                {
+                    Debug.WriteLine($"Failed to invalidate marker: {ex.Message}");
+                }
+            }
+
+            _activeMarkers.Clear(); // Clear the list to avoid references to deleted markers
+        }
+
+
+        private async Task DisplayDiagnosticsAsync(List<CxAscaDetail> scanDetails, string filePath)
+        {
+            if (scanDetails == null || string.IsNullOrEmpty(filePath)) return;
+
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
```

Comment:
why should it be in the main thread?

Reply:
We need to switch to the main thread because we're working with Visual Studio APIs (documents, markers, error list) that require UI thread access. These are COM objects that will throw exceptions if accessed from a different thread.